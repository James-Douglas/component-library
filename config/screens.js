const screens = {
  xs: "375px",
  sm: "720px",
  md: "1200px",
  lg: "1400px",
  xl: "1900px",
  xxl: "2560px"
};

export default screens;
