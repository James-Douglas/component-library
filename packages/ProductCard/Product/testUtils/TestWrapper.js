import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
