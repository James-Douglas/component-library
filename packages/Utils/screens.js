const screens = {
  xs: '375px',
  sm: '768px',
  md: '1024px',
  lg: '1200px',
  xl: '1440px',
};

export default screens;
