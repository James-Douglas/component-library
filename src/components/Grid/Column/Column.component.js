import React from 'react';
import PropTypes from 'prop-types';
import StyledColumn from './Column.styles';

const Column = ({
  cols,
  sm,
  md,
  lg,
  xl,
  xxl,
  offset,
  offsetSm,
  offsetMd,
  offsetLg,
  offsetXl,
  offsetXxl,
  children,
  className,
  valign,
  halign,
}) => (
  <StyledColumn
    cols={cols}
    sm={sm}
    md={md}
    lg={lg}
    xl={xl}
    xxl={xxl}
    baseOffset={offset}
    offsetSm={offsetSm}
    offsetMd={offsetMd}
    offsetLg={offsetLg}
    offsetXl={offsetXl}
    offsetXxl={offsetXxl}
    valign={valign}
    halign={halign}
    className={`
        ${className}
      `}
  >
    {children}
  </StyledColumn>
);

Column.propTypes = {
  /**
   * Classes to be applied to the container element
   */
  className: PropTypes.string,
  /**
   * Width in columns (across all breakpoints where a specific width [sm, md, lg, xl, xxl] has not been given)
   */
  cols: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at sm screen sizes.
   */
  sm: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at md screen sizes.
   */
  md: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at lg screen sizes.
   */
  lg: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at xl screen sizes.
   */
  xl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at xxl screen sizes.
   */
  xxl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns (across all breakpoints where a specific offset [sm, md, lg, xl, xxl] has not been given)
   */
  offset: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the sm screen size.
   */
  offsetSm: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the md screen size.
   */
  offsetMd: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the lg screen size.
   */
  offsetLg: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the xl screen size.
   */
  offsetXl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the xxl screen size.
   */
  offsetXxl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * The column contents.
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  /**
   * The align-items property is related to CSS layout.
   */
  valign: PropTypes.oneOf([
    'flex-start',
    'flex-end',
    'center',
    'baseline',
    'stretch',
  ]),
  /**
   * The justify-content property is related to CSS layout.
   */
  halign: PropTypes.oneOf([
    'flex-start',
    'flex-end',
    'center',
    'space-between',
    'space-around',
    'initial',
    'inherit',
  ]),
};

Column.defaultProps = {
  className: '',
  cols: 0,
  sm: 0,
  md: 0,
  lg: 0,
  xl: 0,
  xxl: 0,
  offset: 0,
  offsetSm: 0,
  offsetMd: 0,
  offsetLg: 0,
  offsetXl: 0,
  offsetXxl: 0,
  children: [],
  valign: 'stretch',
  halign: 'initial',
};

export default Column;
