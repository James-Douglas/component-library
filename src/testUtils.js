// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
