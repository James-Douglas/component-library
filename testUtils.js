import React from 'react';
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './packages/Provider/ManorProvider';

const ManorTestWrapper = ({ children }) => (
  <ManorProvider>
    {children}
  </ManorProvider>
);

const customRender = (ui, options) => render(ui, { wrapper: ManorTestWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
