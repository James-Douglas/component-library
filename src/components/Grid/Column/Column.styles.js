import styled, { css } from 'styled-components';
import screens from '../../../../config/screens';

// for cols='sm' or cols='md' etc
const colDefault = css`
  flex-basis: 0%;
  flex-grow: 1;
  max-width: 100%;
`;
// for sm='auto' or md='auto' etc
const colAuto = css`
  flex: 0 0 auto;
  width: auto;
  max-width: 100%;
`;

const getPercentage = (value) => {
  const parsed = parseInt(value, 10);
  return `${(100 / 12) * parsed}%`;
};

const media = [
  { col: 'sm', minWidth: screens.sm },
  { col: 'md', minWidth: screens.md },
  { col: 'lg', minWidth: screens.lg },
  { col: 'xl', minWidth: screens.xl },
  { col: 'xxl', minWidth: screens.xxl },
];

const StyledColumn = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  padding-left: ${({ theme }) => theme.spacing[16]};
  padding-right: ${({ theme }) => theme.spacing[16]};
  align-items: ${(props) => props.valign};
  justify-content: ${(props) => props.halign};

  ${({ cols, baseOffset }) => {
    let colStyles = '';
    if (parseInt(cols, 10) > 0) {
      colStyles += `
        flex: ${getPercentage(cols)};
        max-width: ${getPercentage(cols)};
      `;
    }

    if (baseOffset) {
      colStyles += `
        margin-left: ${getPercentage(baseOffset)};
      `;
    }

    media.forEach((oneMedia) => {
      if (cols === oneMedia.col) {
        colStyles += `@media (min-width: ${oneMedia.minWidth}) {
          ${colDefault}
        }`;
      }
    });

    return colStyles;
  }}

  @media (min-width: ${screens.xs}) {
    ${({ sm, offsetSm }) => {
    let colStyles = '';

    if (sm === 'auto') {
      return colAuto;
    }

    if (sm) {
      colStyles += `
          flex: ${getPercentage(sm)};
          max-width: ${getPercentage(sm)};
        `;
    }

    if (offsetSm) {
      colStyles += `margin-left: ${getPercentage(offsetSm)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.sm}) {
    ${({ md, offsetMd }) => {
    let colStyles = '';

    if (md === 'auto') {
      return colAuto;
    }

    if (md) {
      colStyles += `
          flex: ${getPercentage(md)};
          max-width: ${getPercentage(md)};
        `;
    }

    if (offsetMd) {
      colStyles += `margin-left: ${getPercentage(offsetMd)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.md}) {
    ${({ lg, offsetLg }) => {
    let colStyles = '';

    if (lg === 'auto') {
      return colAuto;
    }

    if (lg) {
      colStyles += `
          flex: ${getPercentage(lg)};
          max-width: ${getPercentage(lg)};
        `;
    }

    if (offsetLg) {
      colStyles += `margin-left: ${getPercentage(offsetLg)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.lg}) {
    ${({ xl, offsetXl }) => {
    let colStyles = '';

    if (xl === 'auto') {
      return colAuto;
    }

    if (xl) {
      colStyles += `
          flex: ${getPercentage(xl)};
          max-width: ${getPercentage(xl)};
        `;
    }

    if (offsetXl) {
      colStyles += `margin-left: ${getPercentage(offsetXl)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.xl}) {
    ${({ xxl, offsetXxl }) => {
    let colStyles = '';

    if (xxl === 'auto') {
      return colAuto;
    }

    if (xxl) {
      colStyles += `
          flex: ${getPercentage(xxl)};
          max-width: ${getPercentage(xxl)};
        `;
    }

    if (offsetXxl) {
      colStyles += `margin-left: ${getPercentage(offsetXxl)}`;
    }

    return colStyles;
  }}
  }

  ${(props) => !props.cols
    && !props.sm
    && !props.md
    && !props.lg
    && !props.xl
    && !props.xxl
    && css`
      -ms-flex-preferred-size: 0;
      flex-basis: 0%;
      -ms-flex-positive: 1;
      flex-grow: 1;
      max-width: 100%;
    `}
`;

export default StyledColumn;
