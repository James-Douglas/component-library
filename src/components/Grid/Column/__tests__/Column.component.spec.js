import React from 'react';
import { render } from '../../../../testUtils';
import 'jest-styled-components';
import screens from '../../../../../config/screens';
import Column from '../Column.component';

describe('Column', () => {
  it('renders correctly without props', () => {
    const { container } = render(<Column />);
    expect(container).toMatchSnapshot();
  });

  it('renders children', () => {
    const { container } = render(
      <Column>
        <div className="child-content">test</div>
      </Column>,
    );

    const childContent = container.querySelector('.child-content');
    expect(container).toMatchSnapshot();
    expect(childContent).toBeInTheDocument();
  });

  it('adds default styles to col if no props are passed', () => {
    const { container } = render(<Column />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toBeDefined();
    expect(column).toHaveStyleRule('display', 'flex');
    expect(column).toHaveStyleRule('flex-basis', '0%');
    expect(column).toHaveStyleRule('max-width', '100%');
  });

  it('defines col size when supplied', () => {
    const { container } = render(<Column cols={5} />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toBeDefined();
    expect(column).toHaveStyleRule('flex', '41.66666666666667%');
  });

  it('defines an offset when supplied', () => {
    const { container } = render(<Column offset={1} />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toBeDefined();
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%');
  });

  it('resizes on breakpoint', () => {
    const { container } = render(<Column sm={8} md={7} lg={6} xl={5} xxl={4} />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();

    expect(column).toHaveStyleRule('flex', '66.66666666666667%', {
      media: `(min-width:${screens.xs})`,
    });
    expect(column).toHaveStyleRule('flex', '58.333333333333336%', {
      media: `(min-width:${screens.sm})`,
    });
    expect(column).toHaveStyleRule('flex', '50%', {
      media: `(min-width:${screens.md})`,
    });
    expect(column).toHaveStyleRule('flex', '41.66666666666667%', {
      media: `(min-width:${screens.lg})`,
    });
    expect(column).toHaveStyleRule('flex', '33.333333333333336%', {
      media: `(min-width:${screens.xl})`,
    });
  });

  it('resizes offsets on breakpoint', () => {
    const { container } = render(<Column offsetSm={8} offsetMd={7} offsetLg={6} offsetXl={5} offsetXxl={4} />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('margin-left', '66.66666666666667%', {
      media: `(min-width:${screens.xs})`,
    });
    expect(column).toHaveStyleRule('margin-left', '58.333333333333336%', {
      media: `(min-width:${screens.sm})`,
    });
    expect(column).toHaveStyleRule('margin-left', '50%', {
      media: `(min-width:${screens.md})`,
    });
    expect(column).toHaveStyleRule('margin-left', '41.66666666666667%', {
      media: `(min-width:${screens.lg})`,
    });
    expect(column).toHaveStyleRule('margin-left', '33.333333333333336%', {
      media: `(min-width:${screens.xl})`,
    });
  });

  it('stacks at breakpoint sm', () => {
    const { container } = render(<Column cols="sm" />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('flex-basis', '0%', {
      media: `(min-width:${screens.sm})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.sm})`,
    });
  });

  it('stacks at breakpoint md', () => {
    const { container } = render(<Column cols="md" />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('flex-basis', '0%', {
      media: `(min-width:${screens.md})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.md})`,
    });
  });

  it('stacks at breakpoint lg', () => {
    const { container } = render(<Column cols="lg" />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('flex-basis', '0%', {
      media: `(min-width:${screens.lg})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.lg})`,
    });
  });

  it('stacks at breakpoint xl', () => {
    const { container } = render(<Column cols="xl" />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('flex-basis', '0%', {
      media: `(min-width:${screens.xl})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.xl})`,
    });
  });

  it('stacks at breakpoint xxl', () => {
    const { container } = render(<Column cols="xxl" />);

    const column = container.getElementsByTagName('div')[0];

    expect(container).toMatchSnapshot();
    expect(column).toHaveStyleRule('flex-basis', '0%', {
      media: `(min-width:${screens.xxl})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.xxl})`,
    });
  });

  it('allows additional classes to be passed through', () => {
    const { container } = render(<Column className="test-class" />);

    expect(container).toMatchSnapshot();
    const test = container.getElementsByClassName('test-class');
    expect(test).toBeDefined();
  });

  it('allows column be auto ', () => {
    const { container } = render(<Column sm="auto" md="auto" lg="auto" xl="auto" xxl="auto" />);
    const column = container.getElementsByTagName('div')[0];
    expect(column).toHaveStyleRule('flex', '0 0 auto', {
      media: `(min-width:${screens.xs})`,
    });
    expect(column).toHaveStyleRule('width', 'auto', {
      media: `(min-width:${screens.sm})`,
    });
    expect(column).toHaveStyleRule('max-width', '100%', {
      media: `(min-width:${screens.md})`,
    });
    expect(column).toHaveStyleRule('flex', '0 0 auto', {
      media: `(min-width:${screens.lg})`,
    });
    expect(column).toHaveStyleRule('width', 'auto', {
      media: `(min-width:${screens.xl})`,
    });
  });
});
