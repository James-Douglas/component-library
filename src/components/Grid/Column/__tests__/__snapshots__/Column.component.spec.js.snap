// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column adds "col" class by default if no props are supplied 1`] = `
<div>
  <style />
  <div
    class="col "
  />
</div>
`;

exports[`Column allows additional classes to be passed through 1`] = `
<div>
  <style />
  <div
    class="col test-class"
  />
</div>
`;

exports[`Column breakpoint column classes get applied as intended 1`] = `
<div>
  <style />
  <div
    class="col-1 col-sm-1 col-md-1 col-lg-1 col-xl-1 col-xxl-1 offset-1 "
  />
</div>
`;

exports[`Column renders correctly without props 1`] = `
<div>
  <style />
  <div
    class="col "
  />
</div>
`;
